task-maker-async/src/store.rs

```diff
@@ -543,20 +550,59 @@ impl Store for StoreService {
             }
         }
 
-        service
+        let file_set = service
             .file_sets
             .get_mut(&handle_info.data_hash)
             .unwrap()
             .get_mut(&handle_info.variant_hash)
-            .unwrap()
-            .finalized = true;
+            .unwrap();
+
+        file_set.finalized = true;
+        file_set.finalization_waiters.drain(..).for_each(|waiter| {
+            let _ = waiter.send(());
+        });
 
         Ok(FileSetHandle {
             id: handle.id,
             mode: HandleMode::Read,
         })
     }
 
+    async fn wait_until_finalized(
+        self,
+        context: Context,
+        handle: FileSetHandle,
+    ) -> Result<(), Error> {
+        loop {
+            let receiver = {
+                let mut service = self.service.lock().unwrap();
+                let entry = service.validate_and_refresh(handle)?;
+                let (data_hash, variant_hash) = if let Entry::Occupied(file_set_entry) = &entry {
+                    let fs = file_set_entry.get();
+                    (fs.data_hash, fs.variant_hash)
+                } else {
+                    panic!("validate_and_refresh cannot return a non-occupied entry");
+                };
+                let file_set = service
+                    .file_sets
+                    .get_mut(&data_hash)
+                    .and_then(|x| x.get_mut(&variant_hash));
+                let file_set = file_set.unwrap();
```

Comment:
If `file_set` gets ultimately unwrapped, why not unwrapping directly instead of using `and_then`?

Reply:
Eh, this saves me an unwrap, but it doesn't change much after all...